### PR TITLE
Allow unused variables with todo!

### LINKED ISSUE #78136: Don't warn on unused arguments in unimplemented functions
The following code:

```rust
fn foo(arg: i32) {
    todo!()
}

fn main() {
    foo(35);
}
```

([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a35976591f81822c014bfa4e21bd34e8))

Currently emits a warning:

```
warning: unused variable: `arg`
 --> src/main.rs:1:8
  |
1 | fn foo(arg: i32) {
  |        ^^^ help: if this is intentional, prefix it with an underscore: `_arg`
  |
  = note: `#[warn(unused_variables)]` on by default

warning: 1 warning emitted
```

But `todo!()` clearly indicates that the argument is unused neither because it's *intended* to be ignored (in which case it'd be appropriate to rename it to `_arg`), nor because the author *forgot* to use it — but simply because the relevant code to use it has not been written yet.

Since `todo!()` itself does not trigger any warnings (meaning, rustc is fine with trying to compile code with parts that are explicitly *not implemented yet*), it doesn't seem appropriate to warn about unused variables that are "covered" by `todo!()`.

### PR DESCRIPTION
**Pull request by charles-r-earp**
Fixes #78136

As stated in the issue, todo! should imply that the function will be implemented at some point but isn't yet, and therefore the user shouldn't have to either apply the #[allow(unused_variables)] themselves or prefix the arguments with underscores, since they will at some point replace the todo! with an implementation. Unlike unimplemented, todo! is a placeholder, so the suggested prefixing of the arguments with underscores does not make sense because that would imply that those arguments are not going to be used, when in fact they may be used when the function is implemented. Once the todo! is removed, the lint will return if applicable. 

The lint error is emitted in [rustc_passes::liveness::Liveness](https://github.com/rust-lang/rust/blob/master/compiler/rustc_passes/src/liveness.rs)::report_unused. This can be traced back to the \<IrMaps as Visiter>::visit_body method. Unfortunately the Body argument here has been fully expanded, so we cannot match on the todo! macro directly. Instead, the idea is to detect a trailing expression, ie the last expression without a semicolon, that is core::panicking::panic (also std). If the panic is called with "not yet implemented", which is what is emitted by todo!, then we return early and don't do any checking for unused variables. 

For example:

    fn foo(x: i32) -> i32 {
        // some code
        todo!()
    }

### CODE REVIEW

**Comment by jonas-schievink** [comment]
Macros cannot expand to inner attributes at all

**Comment by scottmcm** [comment]
I think this PR should probably have a ui test to confirm that it's working?

**Comment by charles-r-earp** [comment]
> I think this PR should probably have a ui test to confirm that it's working?

You mean a unit test? Where would that go? In core/tests.rs?

**Comment by scottmcm** [comment]
No, a ui test.  See https://rustc-dev-guide.rust-lang.org/tests/adding.html#guide-to-the-ui-tests

File: compiler/rustc_passes/src/liveness.rs [hunk 1]
```diff
@@ -329,6 +329,44 @@ impl<'tcx> Visitor<'tcx> for IrMaps<'tcx> {
...
+                                if let ExprKind::Lit(spanned) = &arg.kind {
+                                    if let LitKind::Str(symbol, StrStyle::Cooked) = spanned.node {
+                                        // FIXME: Better way of matching symbol
+                                        if &*symbol.as_str() == "not yet implemented" {
+                                            return;
+                                        }
+                                    }
+                                }
+                            }
+                        }
```
Comment (camelid) [inline compiler/rustc_passes/src/liveness.rs:364]: This should be a less hacky way of determining if it calls panic:

```suggestion
                        if matches!(&path.res, Res::Def(DefKind::Fn, path_def_id)) {
                            let begin_panic_def_id = self.tcx.lang_items().begin_panic_fn();
                            if begin_panic_def_id == Some(path_def_id) {
                                // FIXME: Better way of extracting literal
                                if let ExprKind::Lit(spanned) = &arg.kind {
                                    if let LitKind::Str(symbol, StrStyle::Cooked) = spanned.node {
                                        // FIXME: Better way of matching symbol
                                        if &*symbol.as_str() == "not yet implemented" {
                                            return;
                                        }
                                    }
                                }
                            }
                        }
```

Comment (camelid) [reply compiler/rustc_passes/src/liveness.rs:364]: However, note that neither the current code nor the code I suggested will catch invocations like `todo!("some message")` that override the panic message.

File: compiler/rustc_passes/src/liveness.rs [hunk 2]
```diff
@@ -329,6 +329,40 @@ impl<'tcx> Visitor<'tcx> for IrMaps<'tcx> {
...
+                todo!()
+            }
+        */
+        if let ExprKind::Block(block, _) = &body.value.kind {
+            if let Some(expr) = block.expr {
+                if let ExprKind::Call(call, [arg]) = expr.kind {
+                    if let ExprKind::Path(QPath::Resolved(_, path)) = call.kind {
+                        if let Res::Def(DefKind::Fn, path_def_id) = &path.res {
+                            let begin_panic_def_id = self.tcx.lang_items().begin_panic_fn();
+                            if begin_panic_def_id == Some(*path_def_id) {
```
Comment (camelid) [inline compiler/rustc_passes/src/liveness.rs:351]: The UI test is failing; I think this is the wrong lang item. Sorry about that! I think we need to change this to catch `core::panicking::panic` and `core::panicking::panic_fmt`, but unfortunately there doesn't appear to be a lang item for `panic_fmt`. So you may need to go back to using string comparisons :/

However, I'm out of my depth here so maybe someone else knows a better way (e.g. **`@m-ou-se`**).

If you do end up needing to use the string comparison approach, you may want to use the [`TyCtxt::def_path_str()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.def_path_str) method to get a path string for `path_def_id`. You would use it like:

```rust
let path_str = self.tcx.def_path_str(path_def_id);
```

File: compiler/rustc_passes/src/liveness.rs [hunk 3]
```diff
@@ -329,6 +329,48 @@ impl<'tcx> Visitor<'tcx> for IrMaps<'tcx> {
             }
         }
 
+        // Allow todo! macro
+        /*
+            Skips checking for unused variables when the trailing expression
+            of the body is a panic with a message that contains "not yet implemented".
```
Comment (camelid) [inline compiler/rustc_passes/src/liveness.rs:335]: I would still include something saying that this is a somewhat hacky way to check if it's `todo!`.

File: compiler/rustc_passes/src/liveness.rs [hunk 4]
```diff
@@ -329,6 +329,48 @@ impl<'tcx> Visitor<'tcx> for IrMaps<'tcx> {
...
+        */
+        if let ExprKind::Block(block, _) = &body.value.kind {
+            if let Some(expr) = block.expr {
+                if let ExprKind::Call(call, [arg]) = expr.kind {
+                    if let ExprKind::Path(QPath::Resolved(_, path)) = call.kind {
+                        if let Res::Def(DefKind::Fn, path_def_id) = &path.res {
+                            let panic_fn = self.tcx.lang_items().panic_fn();
+                            // Note: there is no function for panic_fmt, so we have to extract it from the debug output :(
+                            // builder doesn't like this being called without panic `self.tcx.def_path_str(*path_def_id);`
+                            let path_str = format!("{:?}", path_def_id);
```
Comment (camelid) [inline compiler/rustc_passes/src/liveness.rs:352]: What does this mean? Why did `self.tcx.def_path_str` not work?

File: compiler/rustc_passes/src/liveness.rs [hunk 5]
```diff
@@ -329,6 +329,48 @@ impl<'tcx> Visitor<'tcx> for IrMaps<'tcx> {
...
+                            if Some(*path_def_id) == panic_fn
+                                || ((path_str.contains("std[") || path_str.contains("core["))
```
Comment (camelid) [inline compiler/rustc_passes/src/liveness.rs:354]: By the way, why does this use `[`?

File: compiler/rustc_passes/src/liveness.rs (hunk 4 above)
Comment (charles-r-earp) [reply compiler/rustc_passes/src/liveness.rs:352]: It errors with:
    
    thread 'rustc' panicked at 'no warnings or errors encountered even though `delayed_good_path_bugs` issued', compiler/rustc_errors/src/lib.rs:974:13

File: compiler/rustc_passes/src/liveness.rs (hunk 5 above)
Comment (charles-r-earp) [reply compiler/rustc_passes/src/liveness.rs:354]: The path_str is like `core[76aa]::panicking::panic`. I was concerned about matching on some arbitrary crate with a matching path, though it's a bit unlikely. I actually think it should be `~ core[" .. "]::panicking_panic" ..` just to ensure that it only triggers for std / core.

**Comment by JohnCSimon** [comment]
Ping from triage - @charles-r-earp
can you please resolve the merge conflict? thank you.

@Rustbot label: -S-waiting-on-review +S-waiting-on-author

**Comment by joshtriplett** [comment]
Sorry that I missed the initial ping from rust-highfive here.

I really don't think we should suppress these lint warnings, unless `todo!` itself starts producing a lint warning. Unused variable warnings are a useful *compile-time* indication that something still needs implementing. `todo!` creates a runtime error, so it shouldn't suppress a compile-time warning.

`todo!` is absolutely useful when stubbing out code. That kind of stubbing-out can live for varying lengths of time. A `todo!` might be part of a transient compilation effort to try out an interface, and live for minutes or hours, or it might get committed to the codebase and live for months. In the former situation, I find it useful that the unused variable warnings stick around, as a reminder that I'm not done implementing yet. Modifying variables to use names starting with `_` to suppress those warnings is more of an indication that they'll *stay* unused for an extended period of time, or potentially indefinitely.

So, to me, the usefulness of those warnings goes down *substantially* if `todo!` suppresses them.

**Comment by joshtriplett** [comment]
cc @rust-lang/lang

**Comment by danielhenrymantilla** [comment]
I personally totally agree with @joshtriplett when they say:

> I really don't think we should suppress these lint warnings, unless `todo!` itself starts producing a lint warning.

My personal preference:

 1. Ideally, `todo!`'s would trigger compiler warnings (I have personally written helper macros to achieve this result);
 
 1. Else, the poor man's warning is to make other stuff trigger somewhat related warnings, such as an `unused_variables` warning (Granted, it appearing _multiple times_ if there are multiple such vars is then useless noise).

 1. And finally (what seems to be suggested in this PR), there would be no lint / warning whatsoever.
 
    I do find it quite dangerous, since this has caused code with `todo!`s to end up released to crates.io, out of sheer oversight 😬

**Comment by scottmcm** [comment]
I like the thought of `todo!()` adding a warning instead of the unused variable warning, as one warning for "you have a pending todo" seems more friendly than having a "unused variable" warning for potentially multiple things.

**Comment by nikomatsakis** [comment]
Something I have often found very useful when developing:

Some way to trigger a **warning** during development that would be a **hard error** during ci.

I use this to "make promises" to myself -- but also to my reviewer! -- when doing a patch series. e.g., I might write something broken but include a note that I will get around to fixing it later.

In rustc development, I lean on tidy -- if you write `// FIXME` (I think?) it gives you an error in the tidy service, or it used to. I found this super useful to leave myself notes that I had to come back to before the code could land.

Maybe `todo!` could serve that purpose.

**Comment by camelid** [comment]
> In rustc development, I lean on tidy -- if you write `// FIXME` (I think?) it gives you an error in the tidy service, or it used to. I found this super useful to leave myself notes that I had to come back to before the code could land.

This doesn't seem to be true anymore. I found 1012 FIXMEs in `compiler/`.

**Comment by scottmcm** [comment]
IIRC clippy will allow `// FIXME` but not `// TODO`.

**Comment by nikomatsakis** [comment]
@camelid yeah, my bad, this used to be true; but we do still forbid `// TODO` so that's typically what I use

**Comment by crlf0710** [comment]
Triage: It's not clear what the next steps should be here. From my understanding the next steps would be convert [this comment](https://github.com/rust-lang/rust/pull/79850#issuecomment-816809709) into an issue describing what's the exact language behavior we want, and closing this PR. Could someone on t-lang confirm with this?

**Comment by nikomatsakis** [comment]
I am nominating this PR for @rust-lang/lang team discussion.

**Comment by nikomatsakis** [comment]
We discussed this in our @rust-lang/lang meeting on [2021-05-04](https://github.com/rust-lang/lang-team/blob/master/minutes/2021-05-04.md#allow-unused-variables-with-todo-rust79850). There was definitely a block of us -- maybe even a consensus -- who felt that, early on in the development process, the avalanche of warnings from rustc for "WIP" code could be more annoying than helpful. However, there was generally not a lot of enthusiasm for this *particular* proposal.

One thing we were thinking about is the proposal from [RFC 2383](https://rust-lang.github.io/rfcs/2383-lint-reasons.html#expect-lint-attribute) (tracking issue: https://github.com/rust-lang/rust/issues/54503) to add a `#[expect(lint)]` level. This level would silence the given lint within the given scope -- but report an error if the lint does not fire.

The core idea for this was to enable people to silence "expected" lints for WIP things without accidentally leaving those silences lints in production code (a theme which also arose on this PR).

Based on the conversation, I'm going to go and ahead and move to close, but if folks from @rust-lang/lang feel that's the wrong decision, please speak up. In any case, I'd like to encourage people to consider work on the `#[expect]` idea!

@rfcbot fcp close

**Comment by joshtriplett** [comment]
I'd love to have `expect` exactly as specified in the RFC, and I also like the idea of a version of `expect` that hides all warnings of a given type *and* shows one warning that those warnings were hidden. `expect` would address the case of "this isn't actually an issue in this code, but let me know if the compiler stops complaining about it"; `fixme` would address the case of "this *is* an issue, but it's not the current issue, so keep telling me about it but collapse all such warnings to one".

**Comment by nikomatsakis** [comment]
Going to close. Thanks to all.

**Pull request closed without merging**